### PR TITLE
Maintenance: bump old actions, add dependabot, fix macOS runners configuration

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

Canaries are failing on macOS due to outdated actions. Fixing and preventing future errors.

### Checklist - did you ...

- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -43,7 +43,7 @@ jobs:
             python-version: "3.9"
 
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
```
Comment (kenodegard) [inline .github/workflows/tests.yml:46]: use commit hashes instead?

**Review by marcoesters — COMMENTED** [review]
I'm more familiar with renovate than dependabot, but I have the following questions/suggestions:

* Many actions use only major version numbers. It would be more transparent to add minor and patch numbers, too. In renovate, this also results in better pull request descriptions.
* Will dependabot replace the version numbers with commit hashes eventually? This would result in better security: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

File: .github/workflows/tests.yml (hunk 1 above)
Comment (jaimergp) [reply .github/workflows/tests.yml:46]: We are adding dependabot here to handle that automatically.

Comment (kenodegard) [reply .github/workflows/tests.yml:46]: AFAIK dependabot will replace tags with tags and commits hashes with commits hashes

Comment (jaimergp) [reply .github/workflows/tests.yml:46]: Oh, really? That's unfortunate 😬 Let's see if there's a tool to automatically convert.

Comment (jaimergp) [reply .github/workflows/tests.yml]: Used https://github.com/mheap/pin-github-action and some post-processing. All pinned now.

Comment (marcoesters) [reply .github/workflows/tests.yml]: Do we want to add `<major>.<minor>.<patch>` version numbers to all of them to make it easier to see which version is used exactly?

Comment (jaimergp) [reply .github/workflows/tests.yml]: Ah, I see why now. Cool. Added!

**Review by marcoesters — APPROVED** [review]

**Pull request merged**